proposals/1113_ldap-ad-web-ui/README.md

```diff
@@ -0,0 +1,136 @@
+The "pinniped-cli" client is privileged and as such is trusted to handle a user's credentials
+when authenticating with systems that do not provide an authentication UI (i.e. LDAP).
+
+However, Pinniped is planning to introduce support for dynamic OAuth clients.
+These clients should _not_ be trusted to handle a user's IDP credentials.
+Therefore, we need a mechanism for untrusted clients to acquire Pinniped's downstream tokens while
+leaving the IDP credential handling to the Pinniped supervisor.
+
+## Proposal
+Pinniped must provide a simple login screen in order to support UIs that wish
+to authenticate with the Pinniped Supervisor to gain access to a cluster without
+requiring each app to handle IDP credentials.
+
+### Goals and Non-goals
+
+#### Goals
+* Prevent OAuth clients, other than the Pinniped CLI, from providing credentials via the authorization request
+* Provide a minimal feature set (ie user id, password & submit button only)
+* Provide generalized error messaging for failed logins that do not expose sensitive information (i.e. we should say "invalid username or password"
+  but do not expose whether it's the username or password that's incorrect)
+* Provide information easily allowing a user to identify the screen as belonging to Pinniped and which upstream IdP is being represented (e.g. IdP name)
+* Address basic security concerns for web firms (HTTPS, passwords use a password field, CSRF protection, redirect protection)
+* Prevent LDAP injection attacks
+* Rely on the upstream IdP to address advanced security concerns (brute force protection, username enumeration, etc)
+* Screens are accessible and friendly to screen readers
+* Screens are friendly to password managers
+
+#### Non-goals
+* A rich client (ie the use of javascript)
+* Advanced UI features (e.g. remember me, reveal password). These features are better left to identity providers to implement.
+* Branding & customization beyond the information listed in the goals used to identify the login screen belongs to Pinniped.
+* Supporting SSO integrations
+* Internationalization or localization. The CLI doesn't currently support this either.
+
+### Specification / How it Solves the Use Cases
+
+#### API Changes
+
+The supervisor must accept requests from other clients, as detailed
+in the (todo) proposal for dynamic client registration.
+When a client other than pinniped-cli makes an authorization endpoint request with `response_type=code` and their
+IDP is an LDAP or Active Directory IDP, the user will be redirected to the new login page.
+When a client other than the Pinniped CLI makes an authorization endpoint request with
+custom Username/Password headers, they should be rejected.
+
+The pinniped cli should default to using the cli-based password flow, but when a tty is unavailable or when
+a CLI flag (`--ldap-browser-login`? `--force-browser-login`?) is specified, it will open a browser to log in
```

Comment:
I am unsure if we need a CLI flag for this.

Reply:
Based on https://github.com/vmware-tanzu/pinniped/pull/1116#discussion_r850455917 which says:

> Newer CLIs should fallback to this option when there is no TTY/env vars set. I think more of the flow logic needs to be moved into the login command and out of the get kubeconfig command (to allow this decision to be made at runtime).

I think you're saying that the CLI can make the decision to direct the user to the web UI based on data available at runtime.  Is that right?

Reply:
> I think you're saying that the CLI can make the decision to direct the user to the web UI based on data available at runtime. Is that right?

Yeah, and that the `login oidc` command already has over 20 flags, and I would like to not add more unless we really need to.

Reply:
Resolving this because the existing `--upstream-identity-provider-flow` flag covers this case.

Reply:
Resolved because the existing `upstream-identity-provider-flow` flag covers it